psamm/datasource/native.py

```diff
@@ -174,28 +174,38 @@ class NativeModel(object):
     directory subtree that specifies part of the model.
     """
 
-    def __init__(self, path):
+    def __init__(self, model_from):
         """Create a model from the specified model file or directory"""
-
-        if os.path.isfile(path):
-            self._context = FilePathContext(path)
-            with open(self._context.filepath, 'r') as f:
-                self._model = yaml_load(f)
-        else:
-            # Try to open the default file
-            for filename in DEFAULT_MODEL:
-                try:
-                    self._context = FilePathContext(
-                        os.path.join(path, filename))
-                    with open(self._context.filepath, 'r') as f:
-                        self._model = yaml_load(f)
-                        break
-                except:
-                    logger.debug('Failed to load model file', exc_info=True)
+        if type(model_from) is str:
+            if '\n' in model_from:
+                self._model = yaml_load(model_from)
+                self._context = None
+            elif os.path.isfile(model_from):
+                self._context = FilePathContext(model_from)
+                with open(self._context.filepath, 'r') as f:
+                    self._model = yaml_load(f)
             else:
-                # No model could be loaded
-                raise ParseError('No model file could be found ({})'.format(
-                    ', '.join(DEFAULT_MODEL)))
+                # Try to open the default file
+                for filename in DEFAULT_MODEL:
+                    try:
+                        self._context = FilePathContext(
+                            os.path.join(model_from, filename))
+                        with open(self._context.filepath, 'r') as f:
+                            self._model = yaml_load(f)
+                            break
+                    except:
+                        logger.debug('Failed to load model file',
+                                     exc_info=True)
+                else:
+                    # No model could be loaded
+                    raise ParseError('No model file could be found({})'.format(
+                        ', '.join(DEFAULT_MODEL)))
+        elif type(model_from) is dict:
+            self._context = None
+            self._model = model_from
+        elif type(model_from) is file:
```

Comment:
Usually, it is not a good idea to determine if a variable is a file based on the type. There exists alternative file-like objects that behave exactly like a file but is not `file` type and not even a type that is a subclass to `file` (so `isinstance` does not help either). Instead this should be handled by duck-typing or checking explicitly that an attribute is available. For example, to check whether a variable holds a file-like object that we can read from, we can check that it has a `read` method: `if hasattr(model_from, 'read') and callable(model_from.read):...` or we can try to read from it and catch the exception if it fails (if there is a specific exception for this).